editor/src/clj/editor/debugging/mobdebug.clj

```diff
@@ -109,31 +201,114 @@
+       :default ->LuaBlackBox}
+      edn-string)
+    (catch Exception e
+      (throw (ex-info "Error decoding serialized data" {:input edn-string} e)))))
+
+;; A representation of lua table that supports circular references. Acts like a
+;; read-only map (no assoc/dissoc) that creates/loads sub-structures on demand.
+
+(defn- maybe-ref->structure [^DebugSession debug-session x refs]
+  (if (instance? LuaRef x)
+    (if (contains? refs x)
+      (->LuaStructure debug-session x refs)
+      ;; We have a reference that was not serialized, load it:
+      (with-session debug-session
+        (when (= :suspended (-state debug-session))
+          (let [out (.-out debug-session)
+                in (.-in debug-session)]
+            (send-command! out (format "REF %s --{maxlevel=4}" (:address x)))
+            (let [[status rest] (read-status in)]
+              (case status
+                "200"
+                (when-let [[edn-string] (re-match #"^OK\s+" rest)]
+                  (let [loaded-refs (::refs (decode-serialized-data debug-session edn-string))]
+                    (->LuaStructure debug-session x (into refs loaded-refs))))
+
+                "401"
+                (when-let [[size] (re-match #"^Error in Execution\s+(\d+)$" rest)]
+                  (let [message (read-data in (Integer/parseInt size))]
+                    (log/warn :message "Failed to serialize the ref"
+                              :ref (:address x)
+                              :error-message message)
+                    nil))
+
+                "400"
+                (do (log/warn :message "Couldn't load lua table for a ref: not registered on a server"
+                              :ref (:address x)
+                              :error-message rest)
+                    nil)))))))
+    x))
+
+(deftype LuaStructure [^DebugSession debug-session value refs]
   ILookup
   (valAt [this k]
     (.valAt this k nil))
   (valAt [_ k not-found]
-    (let [lookup-key (if (instance? LuaStructure k)
-                       (.-value ^LuaStructure k)
-                       k)
-          ret (get-in refs [value lookup-key] not-found)]
-      (if (instance? LuaRef ret)
-        (LuaStructure. ret refs)
-        ret)))
+    (case k
+      ::refs refs
+      (let [lookup-key (if (instance? LuaStructure k)
+                         (.-value ^LuaStructure k)
+                         k)
+            ret (get-in refs [value lookup-key] not-found)]
+        (maybe-ref->structure debug-session ret refs))))
 
   Seqable
   (seq [_]
     (seq (map (fn [[k v]]
-                (MapEntry. (if (instance? LuaRef k)
-                             (LuaStructure. k refs)
-                             k)
-                           (if (instance? LuaRef v)
-                             (LuaStructure. v refs)
-                             v)))
+                (MapEntry. (maybe-ref->structure debug-session k refs)
+                           (maybe-ref->structure debug-session v refs)))
```

Comment:
If `maybe-ref->structure` returns `nil` because of an error, should we filter out the resulting `MapEntry` here?

Reply:
Good question. I would say no — for simplicity :) 

`seq` on a LuaStructure is only used in the debugger tree view, and we get an error, the debugger disconnects and that whole view is closed, so maybe it does not matter that much?